bridges/otellogrus/hook.go

```diff
@@ -0,0 +1,252 @@
+// LoggerProvider.
+func WithLevels(l []logrus.Level) Option {
+	return optFunc(func(c config) config {
+		c.levels = l
+		return c
+	})
+}
+
+// NewHook returns a new [Hook] to be used as a [logrus.Hook].
+//
+// If [WithLoggerProvider] is not provided, the returned Hook will use the
+// global LoggerProvider.
+func NewHook(options ...Option) *Hook {
+	cfg := newConfig(options)
+	return &Hook{
+		logger: cfg.logger(),
+		levels: cfg.levels,
+	}
+}
+
+// Hook is a [logrus.Hook] that sends all logging records it receives to
+// OpenTelemetry. See package documentation for how conversions are made.
+type Hook struct {
+	logger log.Logger
+	levels []logrus.Level
+}
+
+// Levels returns the list of log levels we want to be sent to OpenTelemetry.
+func (h *Hook) Levels() []logrus.Level {
+	return h.levels
+}
+
+// Fire handles the passed record, and sends it to OpenTelemetry.
+func (h *Hook) Fire(entry *logrus.Entry) error {
+	ctx := entry.Context
+	h.logger.Emit(ctx, h.convertEntry(entry))
+	return nil
+}
+
+func (h *Hook) convertEntry(e *logrus.Entry) log.Record {
+	var record log.Record
+	record.SetTimestamp(e.Time)
+	record.SetBody(log.StringValue(e.Message))
+
+	const sevOffset = logrus.Level(log.SeverityDebug) - logrus.DebugLevel
+	record.SetSeverity(log.Severity(e.Level + sevOffset))
+	record.AddAttributes(convertFields(e.Data)...)
+
+	return record
+}
+
+func convertFields(fields logrus.Fields) []log.KeyValue {
+	kvs := make([]log.KeyValue, 0, len(fields))
+	for k, v := range fields {
+		kvs = append(kvs, log.KeyValue{
+			Key:   k,
+			Value: convertValue(v),
+		})
+	}
+	return kvs
+}
+
+func convertValue(v interface{}) log.Value {
+	switch v := v.(type) {
+	case bool:
+		return log.BoolValue(v)
+	case []byte:
+		return log.BytesValue(v)
+	case float64:
+		return log.Float64Value(v)
+	case int:
+		return log.IntValue(v)
+	case int64:
+		return log.Int64Value(v)
+	case string:
+		return log.StringValue(v)
+	}
+
+	t := reflect.TypeOf(v)
+	if t == nil {
+		return log.Value{}
+	}
+	val := reflect.ValueOf(v)
+	switch t.Kind() {
+	case reflect.Struct:
+		return log.StringValue(fmt.Sprintf("%+v", v))
+	case reflect.Slice, reflect.Array:
+		items := make([]log.Value, 0, val.Len())
+		for i := 0; i < val.Len(); i++ {
+			items = append(items, convertValue(val.Index(i).Interface()))
+		}
+		return log.SliceValue(items...)
+	case reflect.Map:
+		kvs := make([]log.KeyValue, 0, val.Len())
+		for _, k := range val.MapKeys() {
+			var key string
+			// If the key is a struct, use %+v to print the struct fields.
+			if k.Kind() == reflect.Struct {
+				key = fmt.Sprintf("%+v", k.Interface())
+			} else {
+				key = fmt.Sprintf("%v", k.Interface())
+			}
+			kvs = append(kvs, log.KeyValue{
+				Key:   key,
+				Value: convertValue(val.MapIndex(k).Interface()),
+			})
+		}
+		return log.MapValue(kvs...)
+	case reflect.Ptr, reflect.Interface:
+		if val.IsNil() {
+			return log.Value{}
```

Comment:
not tested

Reply:
I extracted this condition from #5357, but it appears to be redundant for logrus, as we already check for `nil` type earlier.
I've removed it.